x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/AuditMlNotificationAction.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.ml.action;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.core.common.notifications.Level;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Action to create ml notification messages from outside the
+ * ml plugin. Designed for internal use the action is not
+ * exposed via a REST endpoint. Within the ml plugin the Audit
+ * classes should be used directly.
+ */
+public class AuditMlNotificationAction extends ActionType<AcknowledgedResponse> {
+
+    public static final AuditMlNotificationAction INSTANCE = new AuditMlNotificationAction();
+    public static final String NAME = "cluster:internal/xpack/ml/notification";
+
+    private AuditMlNotificationAction() {
+        super(NAME, AcknowledgedResponse::readFrom);
+    }
+
+    public enum AuditType {
+        ANOMALY_DETECTION,
+        DATAFRAME_ANALYTICS,
+        INFERENCE,
+        SYSTEM
+    }
+
+    public static class Request extends ActionRequest {
+        private final AuditType auditType;
+        private final String id;
+        private final String message;
+        private final Level level;
+
+        public Request(AuditType auditType, String id, String message, Level level) {
+            this.auditType = auditType;
+            this.id = id;
+            this.message = message;
```

Comment:
nit: could use `Objects.requireNonNull` here, as the serialization assumes not null